C error when casting `none` to `?int` in map access `or` block

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
V full version: V 0.4.3 5b96d8d.e19b2dd
OS: macos, macOS, 14.2.1, 23C71
Processor: 11 cpus, 64bit, little endian, Apple M3 Pro

vexe mtime: 2024-01-05 00:03:27

vroot: OK
VMODULES: OK
VTMP: OK

Git version: git version 2.39.3 (Apple Git-145)
Git vroot status: weekly.2024.01.test2-1-ge19b2dd4
.git/config present: true

CC version: Apple clang version 15.0.0 (clang-1500.1.0.2.5)
thirdparty/tcc status: thirdparty-macos-arm64 a668e5a0

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg map.v`
```v
fn get_value(m map[string]int, key string) ?int {
	return m[key] or { ?int(none) }
}

```

**What did you expect to see?**

The program to compile successfully.

**What did you see instead?**
```
==================
/tmp/v_501/map.17440313105902365190.tmp.c:12660:20: error: assigning to 'int' from incompatible type '_option_int' (aka 'struct _option_int')
                *(int*) _t2.data = (_option_int){ .state=2, .err=_const_none__, .data={EMPTY_STRUCT_INITIALIZATION} };
                                 ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -1284,7 +1284,7 @@ fn (mut c Checker) check_or_last_stmt(mut stmt ast.Stmt, ret_type ast.Type, expr
 				last_stmt_typ := c.expr(mut stmt.expr)
 
 				if last_stmt_typ.has_flag(.option) || last_stmt_typ == ast.none_type {
-					if stmt.expr in [ast.Ident, ast.SelectorExpr, ast.CallExpr, ast.None] {
+					if stmt.expr in [ast.Ident, ast.SelectorExpr, ast.CallExpr, ast.None, ast.CastExpr] {
 						expected_type_name := c.table.type_to_str(ret_type.clear_flags(.option,
 							.result))
 						got_type_name := c.table.type_to_str(last_stmt_typ)

--- a/vlib/v/checker/tests/map_index_or_block_type_mismatch_err.out
+++ b/vlib/v/checker/tests/map_index_or_block_type_mismatch_err.out
@@ -1,0 +1,13 @@
+vlib/v/checker/tests/map_index_or_block_type_mismatch_err.vv:4:20: error: `or` block must provide a value of type `int`, not `none`
+    2 | fn index_got_none_from_or_block() {
+    3 |     m := map[string]int{}
+    4 |     _ = m['key'] or { none }
+      |                       ~~~~
+    5 |     _ = m['key'] or { ?int(none) }
+    6 | }
+vlib/v/checker/tests/map_index_or_block_type_mismatch_err.vv:5:21: error: `or` block must provide a value of type `int`, not `?int`
+    3 |     m := map[string]int{}
+    4 |     _ = m['key'] or { none }
+    5 |     _ = m['key'] or { ?int(none) }
+      |                        ~~~~~~~~~
+    6 | }

--- a/vlib/v/checker/tests/map_index_or_block_type_mismatch_err.vv
+++ b/vlib/v/checker/tests/map_index_or_block_type_mismatch_err.vv
@@ -1,0 +1,6 @@
+// for issue 20390
+fn index_got_none_from_or_block() {
+	m := map[string]int{}
+	_ = m['key'] or { none }
+	_ = m['key'] or { ?int(none) }
+}

